packages/pyright-internal/src/analyzer/importResolver.ts

```diff
@@ -1310,14 +1324,112 @@ export class ImportResolver {
         }
     }
 
-    private _getStdlibTypeshedPath(execEnv: ExecutionEnvironment, importFailureInfo: string[]) {
-        return this._getTypeshedSubdirectory(/* isStdLib */ true, execEnv, importFailureInfo);
+    // Returns the directory for a module within the stdlib typeshed directory.
+    // If moduleDescriptor is provided, it is filtered based on the VERSIONS
+    // file in the typeshed stubs.
+    private _getStdlibTypeshedPath(
+        execEnv: ExecutionEnvironment,
+        importFailureInfo: string[],
+        moduleDescriptor?: ImportedModuleDescriptor
+    ) {
+        const subdirectory = this._getTypeshedSubdirectory(/* isStdLib */ true, execEnv, importFailureInfo);
+        if (
+            subdirectory &&
+            moduleDescriptor &&
+            !this._isStdlibTypeshedStubValidForVersion(moduleDescriptor, execEnv, importFailureInfo)
+        ) {
+            return undefined;
+        }
+
+        return subdirectory;
     }
 
     private _getThirdPartyTypeshedPath(execEnv: ExecutionEnvironment, importFailureInfo: string[]) {
         return this._getTypeshedSubdirectory(/* isStdLib */ false, execEnv, importFailureInfo);
     }
 
+    private _isStdlibTypeshedStubValidForVersion(
+        moduleDescriptor: ImportedModuleDescriptor,
+        execEnv: ExecutionEnvironment,
+        importFailureInfo: string[]
+    ) {
+        if (!this._cachedTypeshedStdLibModuleVersions) {
+            this._cachedTypeshedStdLibModuleVersions = this._readTypeshedStdLibVersions(execEnv, importFailureInfo);
+        }
+
+        const versionRange = this._cachedTypeshedStdLibModuleVersions!.get(moduleDescriptor.nameParts[0]);
+        if (versionRange) {
+            if (execEnv.pythonVersion < versionRange.min) {
+                return false;
+            }
+
+            if (versionRange.max !== undefined && execEnv.pythonVersion > versionRange.max) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private _readTypeshedStdLibVersions(
+        execEnv: ExecutionEnvironment,
+        importFailureInfo: string[]
+    ): Map<string, SupportedVersionRange> {
+        const versionRangeMap = new Map<string, SupportedVersionRange>();
+
+        // Read the VERSIONS file from typeshed.
+        const typeshedStdLibPath = this._getTypeshedSubdirectory(/* isStdLib */ true, execEnv, importFailureInfo);
+
+        if (typeshedStdLibPath) {
+            const versionsFilePath = combinePaths(typeshedStdLibPath, 'VERSIONS');
+            try {
+                const fileStats = this.fileSystem.statSync(versionsFilePath);
+                if (fileStats.size > 0 && fileStats.size < 256 * 1024) {
+                    const fileContents = this.fileSystem.readFileSync(versionsFilePath, 'utf8');
+                    fileContents.split(/\r?\n/).forEach((line) => {
```

Comment:
Currently, the `VERSIONS` file doesn't support comments, and I think neither mypy nor pytype support them, but maybe it's a good idea to add something like this for future proofing:

```typescript
line = line.split('#')[0];
```

Reply:
mypy does support it: https://github.com/python/mypy/blob/866ee316ded61e1f61207ee33b6f31450493ffa3/mypy/modulefinder.py#L761. I agree that it would be useful to support. Once my typeshed PR is done I'd like to write a slightly more precise specification for the VERSIONS file.

Reply:
Good catch. I've added support for comments.